docs/compiler.md

```diff
@@ -1,57 +1,55 @@
-arbitrary number of arguments of kind `Type` but can **only** accept arguments
-of kind `Type`. Put another way: An expression's kind is the same thing as its
-arity, and only expressions of arity=0 (i.e. expressions of kind `Type`) can be
-applied.
-
-To elaborate on the previous point: A user can define a higher kinded data type,
-such as `data Maybe a = Nothing | Just a` in the schema language. However users
-cannot define a data type which has a higher kinded type as a _parameter_, and
-users must "fully saturate" higher kinded types with well-scoped type variables
-or zero arity `Type`s to pass them as arguments to a `TyCon`. Consequently, a
-definition such as `data HKT f a = HKT (f a)` is not valid because `f :: Type ->
-Type`. Restricting the schema language in this way does reduce its
-expressiveness, but allowing higher kinded type parameters would greatly limit
-(or, at the very least, greatly complicate) our ability to codegen for a variety
-of target languages with type systems less expressive than Haskell.
-
-The schema language supports recursive types, and so the compiler must be
-capable of checking recursive types.
-
-In addition to typechecking, the compiler must perform a special check for
+The primary purpose of the compiler is to typecheck the schemata that users
+define via the _Frontend_. The schemata terms are ADT declarations and their
+types are kinds.
+
+Currently the schema language supports:
+
+ 1. types (such as `Int` or `Bool`),
+
+ 2. and type functions (such as `Maybe` or `Either`).
+
+There are future plans to expand this to Higher Kinded Types (such as `MaybeT`,
+`StateT` etc.) - subject to research into _Codegen_ of such types in the target
+languages. The schema language and the _Compiler_ do support recursive
+types.
+
+Schemata terms must be monomorphically kinded, with polymorphic kinds defaulting
+to monomorphic ones. For example `Phantom a = Phantom` would resolve to the
+monomorphic kind `Type → Type` rather than the polymorphic kind `∀a. a → Type`.
+
+<!-- cstml comments: Is this true? I think it's an Aim not a requirement?
+:todo: -->
+
+In addition to typechecking, the compiler must  perform a special check for
 recursive types: It must validate that a recursive type is inhabited (or
 inhabitable). The purpose of this check is to ensure that any schema which
 passes validation is (in principle) a schema for which type definitions and
 typeclass "instances" (which may be simple functions in languages without
 typeclass support) can be generated. As an example, the compiler should reject
-types such as `data F a = F (F a)`, which is uninhabited (and uninhabitable).
+types such as `data F a = F (F a)`, which is uninhabited.
+
+<!-- cstml comments: Is this true? Seems as though this should fail
+parsing. :todo: -->
 
 (Provisional:) Finally, the compiler should _normalize_ expressions as far as it
 is able to. For example, it may be possible to define a data type in the schema
```

Comment:
I don't think it should be a parse error. You should be able to parens around anything mostly. But this seems like a Frontend concern and it connects with the discussion about `compiler.proto` Ty message and how it currently propagates such constructs all the way to the Compiler. For instance

`data G a = G ((Either) ((Maybe) a) Int)`

Would result in `TyAbs "a" $ TyApp (TyApp Either []) [TyApp (TyApp (TyApp Maybe []) [TyVar a]) [TyRef 'Int']]`

So essentially you have `TyApp SomeTy []` with an empty argument list, which should normalize to `SomeTy`.

If we reformulate the `compiler.proto` Ty to use only binary forms, then this becomes solely a Frontend issue.